src/LanguageServer/Impl/Sources/DocumentationSource.cs

```diff
@@ -13,57 +13,83 @@
 // See the Apache Version 2.0 License for specific language governing
 // permissions and limitations under the License.
 
-using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
 using Microsoft.Python.Analysis;
 using Microsoft.Python.Analysis.Types;
 using Microsoft.Python.Core.Text;
+using Microsoft.Python.Parsing.Ast;
 
 namespace Microsoft.Python.LanguageServer.Sources {
     internal abstract class DocumentationSource {
-        public string GetSignatureString(IPythonFunctionType ft, IPythonType self, out IndexSpan[] parameterSpans, int overloadIndex = 0, string name = null) {
+        public string GetSignatureString(IPythonFunctionType ft, IPythonType self, out (IndexSpan, IParameterInfo)[] parameterSpans, int overloadIndex = 0, string funcName = null) {
+            funcName = funcName ?? ft.Name;
             var o = ft.Overloads[overloadIndex];
 
-            var parameterStrings = GetFunctionParameters(ft, out var parameterNameLengths);
-            var returnDoc = o.GetReturnDocumentation(self);
-            var annString = string.IsNullOrEmpty(returnDoc) ? string.Empty : $" -> {returnDoc}";
+            var spans = new List<(IndexSpan, IParameterInfo)>();
+            var builder = new StringBuilder(funcName);
+            builder.Append('(');
 
-            // Calculate parameter spans
-            parameterSpans = new IndexSpan[parameterStrings.Length];
-            name = name ?? ft.Name;
-            var offset = name.Length + 1;
-            for (var i = 0; i < parameterStrings.Length; i++) {
-                parameterSpans[i] = IndexSpan.FromBounds(offset, offset + parameterNameLengths[i]);
-                offset += parameterStrings[i].Length + 2; // name,<space>
-            }
+            var addComma = false;
+            var addMarker = false;
 
-            var combinedParameterString = string.Join(", ", parameterStrings);
-            return $"{name}({combinedParameterString}){annString}";
-        }
+            foreach (var p in o.Parameters.Skip(ft.IsStatic || ft.IsUnbound ? 0 : 1)) {
+                if (addComma) {
+                    builder.Append(", ");
+                } else {
+                    addComma = true;
+                }
+
+                // Positional only markers are not included in the parameters, so keep track
+                // of whether or not positional only parameters have been seen then add the marker
+                // once the positional only section ends.
+                if (p.Kind == ParameterKind.PositionalOnly) {
+                    addMarker = true;
+                } else if (addMarker && p.Kind != ParameterKind.PositionalOnly) {
```

Comment:
Do we care about `def func(a, b, /)` ?

Reply:
What do you mean by "care"? It's a valid case and needs to be handled, yes. I do that below by checking to see if `addMarker` was left true after the loop.

Reply:
OK